### ISSUE TITLE
replace __generate_6_char_string with adjust_string (was str_for_value)

### ISSUE THREAD

**Issue by markusschloesser** [markusschloesser/MackieC4_P3#55]
I didn't do a deep regression, but I bounced around as much as I ever have with your branch loaded and didn't see any problems until I tried to drop an Auto Filter onto the selected Track.  Barf.  But I noticed a couple of things that may be of interest.

```
2021-04-01T23:59:35.070600: info: Python: INFO:MarkusC4test.MackieC4:69 - (MackieC4) Param 0 name <Frequency>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:69 - (MackieC4) Param 1 name <Resonance>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 2 name <Env. Attack>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 3 name <Env. Release>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 4 name <Env. Modulation>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 5 name <LFO Amount>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 6 name <LFO Frequency>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 7 name <LFO Phase>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 8 name <Filter Type>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 9 name <LFO Quantize On>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 10 name <LFO Quantize Rate>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 11 name <LFO Stereo Mode>
2021-04-01T23:59:35.071600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 12 name <LFO Spin>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:71 - (MackieC4) Param 13 name <LFO Sync>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 14 name <LFO Sync Rate>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 15 name <LFO Offset>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 16 name <>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 17 name <>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 18 name <>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 19 name <>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 20 name <>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 21 name <S/C On>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 22 name <S/C Mix>
2021-04-01T23:59:35.072600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 23 name <S/C Gain>
```
Above is "first pass" logging param names
Below is "second pass" logging the same as tuple elements
```
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 0 tuple name <Frequency>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 1 tuple name <Resonance>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 2 tuple name <Env. Attack>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 3 tuple name <Env. Release>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 4 tuple name <Env. Modulation>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 5 tuple name <LFO Amount>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 6 tuple name <LFO Frequency>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 7 tuple name <LFO Phase>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 8 tuple name <Filter Type>
2021-04-01T23:59:35.073600: info: Python: INFO:MarkusC4test.MackieC4:72 - (MackieC4) Param 9 tuple name <LFO Quantize On>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 10 tuple name <LFO Quantize Rate>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 11 tuple name <LFO Stereo Mode>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 12 tuple name <LFO Spin>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 13 tuple name <LFO Sync>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 14 tuple name <LFO Sync Rate>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 15 tuple name <LFO Offset>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 16 tuple name <>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 17 tuple name <>
2021-04-01T23:59:35.074600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 18 tuple name <>
2021-04-01T23:59:35.075600: info: Python: INFO:MarkusC4test.MackieC4:73 - (MackieC4) Param 19 tuple name <>
2021-04-01T23:59:35.075600: info: Python: INFO:MarkusC4test.MackieC4:75 - (MackieC4) Param 20 tuple name <>
2021-04-01T23:59:35.075600: info: Python: INFO:MarkusC4test.MackieC4:75 - (MackieC4) Param 21 tuple name <S/C On>
2021-04-01T23:59:35.075600: info: Python: INFO:MarkusC4test.MackieC4:75 - (MackieC4) Param 22 tuple name <S/C Mix>
2021-04-01T23:59:35.075600: info: Python: INFO:MarkusC4test.MackieC4:75 - (MackieC4) Param 23 tuple name <S/C Gain>
```
below is "third pass" logging the data received from Live? (to be displayed on the C4?)
Notice the two "illegal character" lines
```
2021-04-01T23:59:35.077600: info: Python: INFO:MarkusC4test.MackieC4:76 - (MackieC4) potIndex<8> cc_value<12.5 kHz> received
2021-04-01T23:59:35.077600: info: Python: INFO:MarkusC4test.MackieC4:76 - (MackieC4) potIndex<9> cc_value<14 %> received
2021-04-01T23:59:35.077600: info: Python: INFO:MarkusC4test.MackieC4:76 - (MackieC4) potIndex<10> cc_value<6.00 ms> received
2021-04-01T23:59:35.077600: info: Python: INFO:MarkusC4test.MackieC4:76 - (MackieC4) potIndex<11> cc_value<200 ms> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<12> cc_value<0.00> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<13> cc_value<0.00> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<14> cc_value<0.11 Hz> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - Logging encountered illegal character(s)!
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<16> cc_value<Lowpass> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<17> cc_value<Off> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<18> cc_value<2> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<19> cc_value<Phase> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<20> cc_value<0.0 %> received
2021-04-01T23:59:35.078600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<21> cc_value<Free> received
2021-04-01T23:59:35.079600: info: Python: INFO:MarkusC4test.MackieC4:78 - (MackieC4) potIndex<22> cc_value<1/16> received
2021-04-01T23:59:35.079600: info: Python: INFO:MarkusC4test.MackieC4:78 - Logging encountered illegal character(s)!
```
Then we see the "boost error", followed by a lot of "display update" vomit.

Skipping past the boost error for a moment, we see the "display update" method is reporting `'ascii' codec can't encode character u'\xb0' in position 4: ordinal not in range(128)`  That's just a more verbose way of saying ` encountered illegal character` over and over again.  So I think the prime suspects are (the data from?) "Auto Filter" Parameters `Param 15 name <LFO Offset>` and `Param 23 name <S/C Gain>`

This is the 00B0 (u'\xb0') unicode character
00B0 ° DEGREE SIGN
• this is a spacing character
→ 02DA ˚  ring above
→ 030A $̊  combining ring above
→ 2070 ⁰  superscript zero
→ 2218 ∘  ring operator

This kind of makes sense maybe if the LFO offset is expressed like phase, 45, 60, 90, 120 degree offsets or whatever.  (any bad "S/C Gain" character won't be encountered until we get past the 00B0 issue)

Now looking at the boost error  `**this**` (second parameter) is what I suspect might be the part that doesn't match.
```
Live.MidiMap.map_midi_cc_with_feedback_map(
    midi_map_handle, 
    **param**,  
    0, 
    encoder, 
    Live.MidiMap.MapMode.relative_signed_bit, 
    feedback_rule, 
    needs_takeover, 
    sensitivity=1.0)

MidiMap.map_midi_cc_with_feedback_map(
int, 
**unicode**, 
int, 
int, 
MapMode, 
CCFeedbackRule, 
bool)


did not match C++ signature:
    
map_midi_cc_with_feedback_map(
	unsigned int midi_map_handle, 
	class TPyHandle<class ATimeableValue> **parameter**, 
	int midi_channel, 
	int controller_number, 
	enum NRemoteMapperTypes::TControllerMapMode map_mode, 
	class NPythonMidiMap::TCCFeedbackRule feedback_rule, 
	bool avoid_takeover, 
	float sensitivity=1.0)
```
Both the first (param) and the third (parameter) signatures agree, but the middle signature says the second parameter is "unicode" which means "string", and a string is definitely not "a timeable value"

Also, `float sensitivity=1.0` above means it is an optional parameter.  If 1.0 sensitivity is what you want, then you don't need to include that parameter when you call the method, it will take the value 1.0 by default.  If you want 0.50 sensitivity, or 5.0 sensitivity, then you need to provide those values when you call the method.

_Originally posted by @BilldarBagdar in https://github.com/markusschloesser/MackieC4_P3/discussions/29#discussioncomment-561212_

**Comment by markusschloesser** [comment]
I think that special characters in DisplayParameter fuck with us, and I think the way to fix that is by using `str_for_value`. 
I've tried to implement this in EncoderDisplaySegment but failed miserably 🙄😂
Have you managed to install the stub that gives you the LOM documentation?

**Comment by BilldarBagdar** [comment]
I didn't install the stub so I guess I'm still missing those docs or whatever, but I can Ctrl+Alt+b into classes, objects, and methods, for example.

I tried too.  I didn't fix anything necessarily.  But I think I should have, [several examples](https://www.javatpoint.com/python-string-encode-method) say it should work.  But it's not working.  I uploaded anyway.  Because it's still an issue with the 00B0 character, but now it's because it can't be encoded to ascii which it can't, but "ignore" (errors) doesn't seem to be working.
```
2021-04-08T01:33:41.931800: info: RemoteScriptError:   File "C:\ProgramData\Ableton\Live 10 Suite\Resources\MIDI Remote Scripts\MarkusC4test\EncoderDisplaySegment.py", line 78, in get_lower_text

2021-04-08T01:33:41.931800: info: RemoteScriptError:     
2021-04-08T01:33:41.931800: info: RemoteScriptError: ascii_encoded = str(self.__lower_text).encode('ascii','ignore')

2021-04-08T01:33:41.931800: info: RemoteScriptError: UnicodeEncodeError
2021-04-08T01:33:41.931800: info: RemoteScriptError: : 
2021-04-08T01:33:41.931800: info: RemoteScriptError: 'ascii' codec can't encode character u'\xb0' in position 4: ordinal not in range(128)
2021-04-08T01:33:41.931800: info: RemoteScriptError: 
```

Maybe you can figure out why 'ignore' doesn't seem to be working.  (everything still seems fine until you try to load the device page for an AutoFilter, and then barf)

**Comment by markusschloesser** [comment]
isn't this the problem?
https://github.com/markusschloesser/MackieC4_P3/blob/7e7128fe1ff1879c52fdaa2daa33013e54884834/wip/MackieC4/EncoderController.py#L1444

**Comment by markusschloesser** [comment]
I don't get vomit, but still an immediate disconnect with the boost error. also tried `utf-7` instead of `ascii`, doesn't make a difference. But I got an idea, will try to visualize it and post here

**Comment by markusschloesser** [comment]
https://github.com/markusschloesser/MackieC4_P3/blob/7e7128fe1ff1879c52fdaa2daa33013e54884834/wip/MackieC4/EncoderController.py#L1250
so when we do a conversion to `str` here, isn't that the problem?
Other code I found in other scripts does this
```
def value_to_string(self, value):

        return self._value_labels[int(not self.value)]
```
so converts back to an `int`
mmh 🤔
I also tried leaving out `__generate_6_char_string` but that just uncenters stuff

**Comment by markusschloesser** [comment]
`lower_string2 += (self.__generate_6_char_string(str(l_alt_text))) if lower_string2.isascii else lower_string2.encode('ascii')` also doesn't work

**Comment by BilldarBagdar** [comment]
It doesn't vomit until you try to load device mode for AutoFilter.

Yes, it's blowing up earlier, whenever "display update" is calling that get_text() method when that B0 character is there.

Try this:  (saw it last night while testing)
```
change:  ascii_encoded = str(self.__lower_text).encode('ascii',errors = 'ignore')

to:   ascii_encoded = str(self.__lower_text.value).encode('ascii',errors = 'ignore')
```
> what does the blue line send back? 

There is a log line right after that "midi map with feedback" method, and you can see text like "-inf db" for the sends logged along with every track selection update, but after this change above, look at what appears in the LCDs... (even though "-inf db" remains logged as the "feedback" value)

My conclusion is: When `self.__lower_text` is a `DeviceParameter` object instead of a string, `str(self.__lower_text)` gives you, for example the text, "-inf db" (from the DeviceParameter class's default to_string() method?); and `str(self.__lower_text.value)` give you the underlying "float value" (in this case).  Both the "float value" and the "pretty string" are coming from Live (your green and red lines).  Currently, I think none of the "blue line" stuff in the other direction (from C4 to Live) may be getting logged right now.  Methods like handle_vpot_rotation() and handle_vpot_click() "handle" the events generated by physical actions on the C4.  

But I suspect you may not be imagining the correct full path.  A "midi event" generated on the device goes into and through Live before it goes through our script and back to the device.  Live has already done what it wants (or is doing what it wants in parallel) with the midi event that goes into our script.  Once it goes into our script, the midi event could have come from anywhere, any other controller or the GUI, and our script just updates the C4 display accordingly.  So before a knob twist is reflected in the LCD, the "midi event" went all the way into Live and came back out through the script to update the display on the C4.  In other words, it may help to think of Live as always being between the script and the device in both directions.

**Comment by markusschloesser** [comment]
> It doesn't vomit until you try to load device mode for AutoFilter.

That's the thing, it does not blow up for me in device mode. Will check later tonight with Live 10, if it's the same for me there. Currently it just throws boost error and disconnects

**Comment by BilldarBagdar** [comment]
Found it.   This should be resolved now,  

I am fairly confident this was the actual problem
```
 Param 16 tuple name <>
 Param 17 tuple name <>
 Param 18 tuple name <>
 Param 19 tuple name <>
 Param 20 tuple name <>
```
It looks like maybe there is a mismatch between the "parameters" in DEVICE_DICT and the parameters AutoFilter actually exposes (now) which makes this assignment `parameter = get_parameter_by_name(self.__chosen_plugin, param_name)` not always successful. We were seeing this particular "boost error" because the code was subsequently doing `parameter = ''` which caused the 'unicode' input value that caused the "boost error"

**Comment by BilldarBagdar** [comment]
> Maybe you can figure out why 'ignore' doesn't seem to be working.

I suspect 'ignore' wasn't working because it was never getting hit.  The whole "non-ascii character range" thing was bubbling up out of the `str()` call in `str(self.__lower_text.value)` apparently `str()` only works correctly when characters are "ascii printable" more-or-less.    Eventually last night I found a "Stack-Overflow" answer that said basically, "stop using `str()` to convert unicode to single-byte text and do it the right way..." and the answer linked [here](https://docs.python.org/3/howto/unicode.html)  which is how/why I ended up just changing `str(self.__lower_text.value)` to `unicode(self.__lower_text.value)` and it started working.

**Comment by BilldarBagdar** [comment]
> That's the thing, it does not blow up for me in device mode. 

I (think I) just discovered that part of the issue I've been seeing when testing is because I have been exclusively using the C4 to change tracks and back, for example, or switch selected devices.  When I used the mouse in the GUI to switch selected devices just now, the "missing device preset name" I've been not-seeing in the channel mode LCD "magically" appeared.

...and that's the thing about testing and/or application-users in general, users are very unpredictable.  We might both start from the exact same .als file; but unless we are carefully following the same "regression test script" our unique sequences of actions once the session finishes loading will quickly diverge.  That's not necessarily good or bad, each approach has strengths.

**Comment by markusschloesser** [comment]
> > Maybe you can figure out why 'ignore' doesn't seem to be working.
> 
> I suspect 'ignore' wasn't working because it was never getting hit. The whole "non-ascii character range" thing was bubbling up out of the `str()` call in `str(self.__lower_text.value)` apparently `str()` only works correctly when characters are "ascii printable" more-or-less. Eventually last night I found a "Stack-Overflow" answer that said basically, "stop using `str()` to convert unicode to single-byte text and do it the right way..." and the answer linked [here](https://docs.python.org/3/howto/unicode.html) which is how/why I ended up just changing `str(self.__lower_text.value)` to `unicode(self.__lower_text.value)` and it started working.

i build in a junction(?) to support both py2 and py3, cos py3 can't do unicode. I think, and reading "how to do str and unicode for both version, that quite a bit of different behavior we have seen in the past, is because of this. (https://portingguide.readthedocs.io/en/latest/strings.html)

**Comment by markusschloesser** [comment]
> Found it. This should be resolved now,
> 
> I am fairly confident this was the actual problem
> 
> ```
>  Param 16 tuple name <>
>  Param 17 tuple name <>
>  Param 18 tuple name <>
>  Param 19 tuple name <>
>  Param 20 tuple name <>
> ```
> 
> It looks like maybe there is a mismatch between the "parameters" in DEVICE_DICT and the parameters AutoFilter actually exposes (now) which makes this assignment `parameter = get_parameter_by_name(self.__chosen_plugin, param_name)` not always successful. We were seeing this particular "boost error" because the code was subsequently doing `parameter = ''` which caused the 'unicode' input value that caused the "boost error"

doesn't work for me. It doesn't disconnect anymore, but the 4 parameters now show moooo ;-)

**Comment by BilldarBagdar** [comment]
What doesn't work for you?  You've closed a couple issues, so something must be better.

I was/am getting ready to push an update related to how I wasn't seeing the device name I was expecting to see on the saved-selected-track (after switch track and back in channel mode). 

Empty "send encoder display segments" will show "mooooo cowcow", empty "device encoder display segments" will show " " (and  empty "device parameter encoder display segments" will show "cwsays mooooo") until somebody gets tired of seeing that... :)

**Comment by markusschloesser** [comment]
> What doesn't work for you? You've closed a couple issues, so something must be better.

The auto filter params mentioned here https://github.com/markusschloesser/MackieC4_P3/issues/57#issuecomment-817210997

> I was/am getting ready to push an update related to how I wasn't seeing the device name I was expecting to see on the saved-selected-track (after switch track and back in channel mode).

not sure I follow? 
Is that an existing issue here?

> Empty "send encoder display segments" will show "mooooo cowcow", empty "device encoder display segments" will show " " (and empty "device parameter encoder display segments" will show "cwsays mooooo") until somebody gets tired of seeing that... :)

I am soo tired of them 🙄😂, but I want both lines empty and that was a problem in the past. So haven't done anything

**Comment by BilldarBagdar** [comment]
Ok - I pulled down your last two commits, and then pushed up two of my own.  One of mine is about those "autofilter params".  Those 5 weren't working for me either, until this commit. But I'm not sure, populating them is going to help.  The other commit of mine is about "missing" device names in "channel mode".  That initial "channel mode" display is still basically empty, but now I can change tracks and back, and the loaded device name appears along with everything else.

**Comment by markusschloesser** [comment]
> The other commit of mine is about "missing" device names in "channel mode". That initial "channel mode" display is still basically empty, but now I can change tracks and back, and the loaded device name appears along with everything else.

is it this? https://github.com/markusschloesser/MackieC4_P3/issues/56

**Comment by markusschloesser** [comment]
in general this seems to work but after undoing a device insertion, this happened:

```
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 291, in build_midi_map
RemoteScriptError: s.build_midi_map(midi_map_handle)
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\Encoders.py", line 85, in build_midi_map
RemoteScriptError: Live.MidiMap.map_midi_cc_with_feedback_map(midi_map_handle, param, 0, encoder, Live.MidiMap.MapMode.relative_signed_bit, feedback_rule, needs_takeover, sensitivity=1.0)  # MS "sensitivity" added
RemoteScriptError: Boost.Python
RemoteScriptError: ArgumentError
RemoteScriptError: Python argument types in
    MidiMap.map_midi_cc_with_feedback_map(int, DeviceParameter, int, int, MapMode, CCFeedbackRule, bool)
did not match C++ signature:
    map_midi_cc_with_feedback_map(unsigned int midi_map_handle, class TPyHandle<class ATimeableValue> parameter, int midi_channel, int controller_number, enum NRemoteMapperTypes::TControllerMapMode map_mode, class NPythonMidiMap::TCCFeedbackRule feedback_rule, bool avoid_takeover, float sensitivity=1.0)
Python: INFO:MackieC4.MackieC4:643 - (MackieC4) Not enough devices loaded and __chosen_device is not None: name display blank over device index 0
```
then this vomited all over the place:
```
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\EncoderDisplaySegment.py", line 76, in get_lower_text
RemoteScriptError: ascii_encoded = self.__lower_text.encode('ascii', 'ignore')
RemoteScriptError: AttributeError
RemoteScriptError: 'DeviceParameter' object has no attribute 'encode'
During handling of the above exception, another exception occurred:
RemoteScriptError: Traceback (most recent call last):
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\EncoderDisplaySegment.py", line 79, in get_lower_text
RemoteScriptError: ascii_encoded = unicode(self.__lower_text).encode('ascii', errors='ignore')  # for Py2/Live10
RemoteScriptError: NameError
RemoteScriptError: name 'unicode' is not defined
During handling of the above exception, another exception occurred:
RemoteScriptError: Traceback (most recent call last):
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\MackieC4.py", line 182, in update_display
RemoteScriptError: c.on_update_display_timer()
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\EncoderController.py", line 1437, in on_update_display_timer
RemoteScriptError: l_alt_text = text_for_display.get_lower_text()
RemoteScriptError:   File "C:\ProgramData\Ableton\Live 11 Beta\Resources\MIDI Remote Scripts\MackieC4\EncoderDisplaySegment.py", line 81, in get_lower_text
RemoteScriptError: ascii_encoded = str(self.__lower_text).encode('ascii', errors='ignore')  # for Py3/Live11
RemoteScriptError: Boost.Python
RemoteScriptError: ArgumentError
RemoteScriptError: Python argument types in
    DeviceParameter.__str__(DeviceParameter)
did not match C++ signature:
    __str__(class TPyHandle<class ATimeableValue>)

```

**Comment by markusschloesser** [comment]
so why is deleting a device ok, but undoing device insertion throws up so massively? 🤔

**Comment by BilldarBagdar** [comment]
By "Undoing" do you mean Live's "undo feature"?

`Not enough devices loaded and __chosen_device is not None: name display blank over device index 0`  

This happens when, for example, a device is deleted, but the script doesn't get "cleaned up" properly.  In this case, "on display update" is saying, "There is a "device" object in the `self.__chosen_plugin` variable, but Live's list of devices on the track is empty, so there won't be a "device name" displayed."  But under the hood, we know the device was deleted using Live's "undo feature".  

I suspect there could be an "undo listener" we need to add or fix or something.  The other vomit is probably just GIGO, the script was already off the rails and who knows what the situation was. But...

Instead of a second "try/except"  (catching a NameError) under except: in get_lower_text()
```
        try:
            ascii_encoded = self.__lower_text.encode('ascii', 'ignore')
        except AttributeError:  # DeviceParameter
            ascii_encoded = unicode(self.__lower_text).encode('ascii', errors='ignore')  # for Py2/Live10
```
try importing 'unicode' in that class
```

if sys.version_info[0] >= 3:  # Live 11
    from builtins import range
    from builtins import unicode
```

**Comment by BilldarBagdar** [comment]
> is it this? #56

I don't remember the timing specifics, but what I was talking about was the 6 char lower text device name on a track that was selected by default when the session loads, and the (15 char) "selected device name" in the top LCD.  Both of them are implemented now.  I think I did cause that problem with device and parameter selections hanging up on my way to solving this one though.

**Comment by markusschloesser** [comment]
> By "Undoing" do you mean Live's "undo feature"?

yes. As I have now implemented undo and redo in Function mode, this bug will probably fuck with us a little more often.

> I suspect there could be an "undo listener" we need to add or fix or something.

Don't think so, couldn't find any in the other Mackie scripts. I think it's our "db" that's coming back to haunt us 😶👻
I have focused on "what can I currently achieve with my limited abilities" in the last couple of days, which was ok/nice, even though I know that the code won't win any medals. But hey, at least I am writing actual code 😊
(all commits in shared-dev)

> Instead of a second "try/except" (catching a NameError) under except: in get_lower_text()
try importing 'unicode' in that class

I solved it similarly yesterday

**Comment by markusschloesser** [comment]
done in https://github.com/markusschloesser/MackieC4_P3/commit/9aa017c012804f670e3ed39c6cdaff6ed72ad7b6 
also of course for 15 and 20 char str
now using v2 framework from from `ableton.v2.control_surface.elements.display_data_source import adjust_string`
